File: src/main/java/com/gmail/borlandlp/minigamesdtools/gui/hotbar/items/SlotItem.java
package com.gmail.borlandlp.minigamesdtools.gui.hotbar.items;

import org.bukkit.Sound;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import java.time.Instant;

public abstract class SlotItem {
    protected Long cooldownTime;
    protected ItemStack activeIcon;
    protected ItemStack unactiveIcon;
    protected int amount;
    protected String ID;
    protected long lastClickTime;
    protected String name;
    protected boolean infiniteSlot = false;
    protected Sound useSound;

    public String getID() {
        return ID;
    }

    public void setID(String ID) {
        this.ID = ID;
    }

    public void setActiveIcon(ItemStack activeIcon) {
        this.activeIcon = activeIcon;
    }

    public void setUnactiveIcon(ItemStack unactiveIcon) {
        this.unactiveIcon = unactiveIcon;
    }

    public long getCooldownTime() {
        return cooldownTime;
    }

    public void setCooldownTime(long cooldownTime) {
        this.cooldownTime = cooldownTime;
    }

    public boolean isInfiniteSlot() {
        return infiniteSlot;
    }

    public void setInfiniteSlot(boolean infiniteSlot) {
        this.infiniteSlot = infiniteSlot;
    }

    public int getAmount() {
        return this.amount;
    }

    public void setAmount(int amount) {
        this.amount = amount;
    }

    public ItemStack getUnactiveIcon() {
        return this.unactiveIcon;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public ItemStack getActiveIcon() {
        return this.activeIcon;
    }

    public long getLastClickTime() {
        return lastClickTime;
    }

    public void setLastClickTime(long lastClickTime) {
        this.lastClickTime = lastClickTime;
    }

    public Sound getUseSound() {
        return useSound;
    }

    public void setUseSound(Sound useSound) {
        this.useSound = useSound;
    }

    public long getCooldownRemain() {
        return this.getLastClickTime() == 0L ? 0L : this.getCooldownTime() - ((System.nanoTime() - this.getLastClickTime()) / 1000000);
    }

    public boolean inCooldown() {
        return this.getLastClickTime() != 0L && this.getCooldownRemain() > 0L;
    }

    public void performClick(Player player) {
        if(!this.inCooldown()) {
            this.setLastClickTime(System.nanoTime());
            boolean result = this.use(player);
            if(result) {
                if(!this.isInfiniteSlot()) {
                    this.setAmount(this.getAmount() - 1);
                }
                if(this.getUseSound() != null) {
                    player.getWorld().playSound(player.getLocation(), this.getUseSound(), 1f, 1f);
                }
            }
        } else {
            player.sendMessage("DBG:CD_REMAIN->" + (this.getCooldownRemain()));//DEBUG
        }
    }

    public ItemStack getDrawData() {
        ItemStack icon;
        long msDiff = (System.nanoTime() - this.getLastClickTime()) / 1000000;
        if(msDiff > this.getCooldownTime()) {
            icon = this.getActiveIcon().clone();
            icon.setAmount(this.getAmount());
        } else {//item in cooldown
            icon = this.getUnactiveIcon().clone();
        }

        return icon;
    }

    public abstract boolean use(Player player);

    public String toString() {
        StringBuilder str = new StringBuilder()
                .append("{")
                .append("activeIcon=").append(this.getActiveIcon().getType().name())
                .append(", unactiveIcon=").append(this.getUnactiveIcon().getType().name())
                .append(", amount=").append(this.getAmount())
                .append(", infiniteSlot=").append(this.isInfiniteSlot())
                .append(", name=").append(this.getName())
                .append("}");

        return str.toString();
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/gui/hotbar/items/SlotItem.kt
package com.gmail.borlandlp.minigamesdtools.gui.hotbar.items

import org.bukkit.Sound
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

abstract class SlotItem {
    protected var cooldownTime: Long? = null
    var activeIcon: ItemStack? = null
    var unactiveIcon: ItemStack? = null
    var amount = 0
    var iD: String? = null
    var lastClickTime: Long = 0
    var name: String? = null
    var isInfiniteSlot = false
    var useSound: Sound? = null

    fun getCooldownTime(): Long {
        return cooldownTime!!
    }

    fun setCooldownTime(cooldownTime: Long) {
        this.cooldownTime = cooldownTime
    }

    val cooldownRemain: Long
        get() = if (lastClickTime == 0L) 0L else getCooldownTime() - (System.nanoTime() - lastClickTime) / 1000000

    fun inCooldown(): Boolean {
        return lastClickTime != 0L && cooldownRemain > 0L
    }

    fun performClick(player: Player) {
        if (!inCooldown()) {
            lastClickTime = System.nanoTime()
            val result = this.use(player)
            if (result) {
                if (!isInfiniteSlot) {
                    amount = amount - 1
                }
                if (useSound != null) {
                    player.world.playSound(player.location, useSound, 1f, 1f)
                }
            }
        } else {
            player.sendMessage("DBG:CD_REMAIN->" + cooldownRemain) //DEBUG
        }
    }

    //item in cooldown
    val drawData: ItemStack
        get() {
            val icon: ItemStack
            val msDiff = (System.nanoTime() - lastClickTime) / 1000000
            if (msDiff > getCooldownTime()) {
                icon = activeIcon!!.clone()
                icon.amount = amount
            } else { //item in cooldown
                icon = unactiveIcon!!.clone()
            }
            return icon
        }

    abstract fun use(player: Player): Boolean

    override fun toString(): String {
        val str = StringBuilder()
            .append("{")
            .append("activeIcon=").append(activeIcon!!.type.name)
            .append(", unactiveIcon=").append(unactiveIcon!!.type.name)
            .append(", amount=").append(amount)
            .append(", infiniteSlot=").append(isInfiniteSlot)
            .append(", name=").append(name)
            .append("}")
        return str.toString()
    }
}